store/stat_store.go

```diff
@@ -0,0 +1,323 @@
+		n := uint64(ss.lastPut.actualCount)
+		ss.lastPut.average = float64((newVal + (n * oldAvg)) / (n + 1))
+		ss.lastPut.actualCount++
+		return nil
+	}
+
+	if ts.Before(lastPutTime) {
+		// Ignore TimePoints with Timestamps in the past
+		return fmt.Errorf("the provided timepoint has a timestamp in the past")
+	}
+
+	// Create a new bucket if the buffer is empty
+	if ss.buffer.Front() == nil {
+		ss.newBucket()
+		ss.resetLastPut(ts, tp.Value.(uint64))
+		return nil
+	}
+
+	// Delete the earliest represented TimePoint if the window is full
+	if ss.tpCount == ss.windowDuration {
+		rewind()
+	}
+
+	lastEntry := ss.buffer.Front().Value.(tpEntry)
+
+	// Place lastPut in the latest bucket if the difference from its average
+	// is less than epsilon
+	if Abs(float64(bucketVal-ss.lastPut.average)) < ss.epsilon {
+		lastEntry.count++
+		// TODO: cumulative moving average instead of last two-values? maybe standardize to bucketVal?
+		lastEntry.average = uint64((lastEntry.average + ss.lastPut.average) / 2)
+		if ss.lastPut.max > lastEntry.max {
+			lastEntry.max = ss.lastPut.max
+			lastEntry.maxIdx = lastEntry.count - 1
+		}
+	} else {
+		// Create a new bucket
+		ss.newBucket()
+		ss.resetLastPut(ts, tp.Value.(uint64))
+	}
+	return nil
+
+}
+
+// resetLastPut initializes the lastPut field of the statStore from a time and a value.
+func (ss *statStore) resetLastPut(timestamp time.Time, value uint64) {
+	ss.lastPut.stamp = timestamp
+	ss.lastPut.actualCount = 1
+	ss.lastPut.average = value
+}
+
+// newBucket appends a new bucket to the statStore, using the values of lastPut.
+func (ss *statStore) newBucket() {
+	newEntry := tpEntry{
+		count:   1,
+		average: ss.lastPut.average,
+		max:     ss.lastPut.max,
+		maxIdx:  0,
+	}
+	ss.buffer.PushFront(newEntry)
+
+	// If this was the first bucket, update ss.start
+	if ss.start.Equal(time.Time{}) {
+		ss.start = ss.lastPut.stamp.Add(-ss.resolution)
+	}
+}
+
+// rewind deletes the oldest one resolution of data in the statStore.
+func (ss *statStore) rewind() {
+	firstEntry := ss.buffer.Back().Value.(tpEntry)
+	// Decrement number of TimePoints in the earliest tpEntry
+	firstEntry.count--
+	// Decrement total number of TimePoints in the statStore
+	ss.tpCount--
+	if firstEntry.maxIdx == 0 {
+		// The Max value was just removed, lose precision for other maxes in the entry
+		firstEntry.max = firstEntry.average
+		firstEntry.maxIdx = firstEntry.count - 1
+	} else {
+		firstEntry.maxIdx--
+	}
+	if firstEntry.count == 0 {
+		// Delete the entry if no TimePoints are represented any more
+		ss.buffer.Remove(firstEntry)
+	}
+
+	// Update the start time of the statStore
+	ss.start.Add(ss.resolution)
+}
+
+// Get generates a []TimePoint from the appropriate tpEntries.
+// Get receives a start and end time as inputs.
+func (ss *statStore) Get(start, end time.Time) []TimePoint {
+	ss.RLock()
+	defer ss.RUnlock()
+	if ss.buffer.Len() == 0 {
+		return nil
+	}
+
+	zeroTime := time.Time{}
+	result := []TimePoint{}
```

Comment:
nit: use `var result []TimePoint` to avoid allocation if no elements are to be added